src/lem/circuit.rs

```diff
@@ -0,0 +1,927 @@
+                                        &mut cs.namespace(|| {
+                                            format!(
+                                                "implies equal for {name}'s hash (LEMOP {:?}, pos {i})",
+                                                &op
+                                            )
+                                        }),
+                                        &concrete_path,
+                                        allocated_ptr.hash(),
+                                        &preallocated_preimg[ptr_idx + 1], // hash index
+                                    )?;
+                                }
+                            };
+                        }
+                        macro_rules! hash_helper {
+                            ( $img: expr, $tag: expr, $preimg: expr, $slot: expr ) => {
+                                // Allocate image. Its hash field is constrained by `constrain_slot!`
+                                let allocated_img = LEM::allocate_ptr(
+                                    cs,
+                                    &$img.to_zptr(&g.frame, g.store)?,
+                                    $img.name(),
+                                    &mut g.allocated_ptrs,
+                                )?;
+                                // Create constraint for the tag
+                                let allocated_tag =
+                                    g.global_allocator.get_or_alloc_const(cs, $tag.to_field())?;
+                                implies_equal(
+                                    &mut cs
+                                        .namespace(|| format!("implies equal for {}'s tag", $img)),
+                                    &concrete_path,
+                                    allocated_img.tag(),
+                                    &allocated_tag,
+                                )?;
+
+                                // Retrieve allocated preimage
+                                let allocated_preimg =
+                                    LEM::get_allocated_preimg($preimg, &g.allocated_ptrs)?;
+
+                                // Add the hash constraints
+                                constrain_slot!(
+                                    $preimg,
+                                    $img,
+                                    allocated_preimg,
+                                    allocated_img,
+                                    $slot
+                                );
+                            };
+                        }
+
+                        macro_rules! unhash_helper {
+                            ( $preimg: expr, $img: expr, $slot: expr ) => {
+                                // Allocate preimage to be constrained later by `constrain_slot!`
+                                let allocated_preimg =
+                                    LEM::allocate_preimg(cs, $preimg, &g.frame, g.store, &mut g.allocated_ptrs)?;
```

Comment:
I think these are duplicate allocations because the preimage elements will all have already been allocated in the slot.